broadinstitute/single_cell_portal_core
# Adding custom colors for grouped scatter plots (SCP-2943)

SCP-2943.  This ended up going much faster than expected, so this is, IMHO, actually mergeable code, rather than the prototype suggested by the ticket.  It may not be fully-featured (e.g. we might want a 'reset all to default' button somewhere, or better help text and/or validation), but I would be comfortable releasing this in its current state to get the initial user feedback

If a user has edit permissions for a study, they will see a 'customize colors' option over the legend.  
![image](https://user-images.githubusercontent.com/2800795/151226579-dbe9ab9e-acf0-4ca6-83dd-c45eb6229c43.png)
Clicking customize colors puts the legend into color-editing mode (so you can no longer show/hide traces). 
![image](https://user-images.githubusercontent.com/2800795/151226714-28c720b4-205d-4859-ae41-d6e9dd46207a.png)
Then clicking a legend entry brings up a color picker
![image](https://user-images.githubusercontent.com/2800795/151226795-492f2bcb-9c6d-460b-901c-266422182bc0.png)
The user can then hit 'Save colors' to save any colors they picked to the server.  Once saved, the colors will be visible to all users.

TO TEST:
1.  Open a study for which you do not have edit permissions
2. confirm scatter plot legend appears as before, and you cannot customize colors
3. Open a study for which you do have edit permissions
4. On the explore tab for a cluster, click 'customize colors'
5. Click one or more legend entries, and change the colors
6. click 'save colors'
7. refresh the page, confirm new colors still appear
8. sign out, and view the study, confirm new colors still appear
